### DISCUSSION TITLE
Cannot create a mac application with adt on mac the way a used to do

### DISCUSSION THREAD

**Discussion by danielkotzer** [airsdk/Adobe-Runtime-Support#1127] [General]
I used to create mac applications with adt on my mac with this terminal command:
`bash adt -package -target bundle name.app name.air`

not working anymore get an error: failed to load the air runtime

Even though I  did this: sudo spctl --master-disable to let the mac run applications outside of the store

and this: sudo xattr -rd com.apple.quarantine /the location of the adt file

and even tried this: [https://www.soccertutor.com/tacticsmanager/Resolve-Adobe-AIR-Error-on-MacOS-Catalina.pdf](https://www.soccertutor.com/tacticsmanager/Resolve-Adobe-AIR-Error-on-MacOS-Catalina.pdf)

I even tried to create the mac app from the air file on my windows and then copy it to my mac, then the app is not recognized at all, if I run an old version of this app I did a year ago, on my old computer, it runs perfectly even on my big sur operating system.

I need to be able to create mac apps with the mac air sdk from air files, what should I do?

**Comment by danielkotzer** [comment]
I used my old Mac mini 2011 with high sierra os Air SDK 33.1 and java developer sdk 1.8.0 112 to create a mac app from air, worked fine even on my big sur OS on my Mac mini 2015, but I don't understand who come it is running big sure? you can't sign an air app with with a Developer ID certificate and therefore you can't notarize it, so it is supposed to be blocked by the Gatekeeper, maybe Apple made a special exception for Air? maybe it is one of the settings I made as mentioned above? And why is there no simple app on mac that is helping you to sign a mac app created with Air SDK? I hope someone from Harman is listening... you are doing a great job but how are we suppose to publish a macOS application? can't anyone create an external application like iResign is doing for ipa files, that can resign a mac app created by Air? a simple codesign is not good, because of all the sub applications that all need to be dealt with... what are we supposed to do?

**Reply by ajwfrost** [reply]
Hi - can I clarify what you're trying and what bit doesn't work? Your original post says you did:
```
bash adt -package -target bundle name.app name.air
```
So did that command fail with the error mentioned? Or did you get a "name.app" generated, but then it failed to load this?
What macOS version (and machine/cpu) were you using for this?
And can you confirm the exact version of the AIR SDK (e.g. via `adt -version`)?

For the comment above: so this creates a .app successfully and it also runs on Big Sur, what version of Big Sur do you have there?

There is no exception made by Apple for AIR apps, but we are seeing some odd/inconsistent results sometimes. Originally we stripped the (original Harman) code-signing from generated AIR apps because this was causing a problem, but an unsigned version would work. But then we are finding on some machines that unsigned versions are now being blocked. It may depend both on the exact version number of the OS as well as the user options..

The hopefully-foolproof method to create a properly signed macOS application is to use the latest AIR SDK and package the application directly, with a captive/bundled runtime:
```
adt -package (AIR signing options) -target bundle -storetype KeychainStore -alias "Developer ID Application" name.app app.xml name.swf etc
```
but we were planning to add the same option to the "create from .air" mechanism, hence in the future it should be possible to do:
```
adt -package -target bundle -storetype KeychainStore -alias "Developer ID Application" name.app name.air
```

Under the hood, this is just doing the code-signing using the native codesign command and searching for all the frameworks/binaries that are required to be signed, since the `deep` option doesn't do this for you. You can find out if they're all signed just by trying to notarize it and checking the logs..  We wondered about adding the notarization step too, but that is relatively straightforward to do separately...

thanks

**Comment by danielkotzer** [comment]
I have an 2 Mac mini computers, one from 2011 with OS High Sierra. and one from 2015 with the latest big sur version.
In the past I used the old Mac mini to create a mac application from air file, with no signing, with this command:
`bash adt -package -target bundle name.app name.air`
this was done with Air 33.1.68, and had no problems because I didn't need any signing back there, or it wasn't such a problem distributing an unsigned Mac application.

Lately 2 things have changed:
1 I bought the second Mac mini with newer OS and wanted to created Mac apps from air files with it,
2 I wanted to start signing my applications because it is a real problem distributing unsigned apps

first I accoutered problems creating the mac on my new system, big sur and Air sdk 33.1.1.554
it gave me an error message that no access is aloud to the adt
I thought big sure probably blocks it as it blocks every app that didn't come from the store, so I tried to unlock the adt file
then I got a different message: Cant run the runtime adt or something like this
So I gave up on creating new Mac apps with my new Mac mini and had gone back to my old one, installed an old version of Air sdk 33.1.68 and created my air file on Windows Flash Builder with the same Air sdk. On my Flash Builder, the Jer version is 112, because newer versions made problems with the old certificate I was using, so I installed Java sdk version u8 112 on my mac to make it fit, so there will be no problems, I think in the past I did encounter some problems with newer versions of java sdk.
I created a new mac application, as I was used to, and it worked fine.
Strange thing is that it worked fine even on big sur and I wasn't sure why it didn't block it.
But I wanted to sign it properly,  tried to do it on mac got and error: no native signing is possible on mac.
I could have created a signed mac.app application on Windows from air, but this gave me a file that was not working.
So now you say signing (if I understood you correctly ) - it is not possible to create a signed mac application from air, I can only create it from a swf file with the command you showed above, 
`adt -package (AIR signing options) -target bundle -storetype KeychainStore -alias "Developer ID Application" name.app app.xml name.swf etc
`
OK, I will try it, and this is working on a mac you say?

will this type of signing work on my old High sierra OS or do I need something new?
Because if I need the newest version of macOS and the new Air sdk, they don't wok on my new computer, and I don't know if they will work on my older one.

**Reply by ajwfrost** [reply]
Hi

> But I wanted to sign it properly, tried to do it on mac got and error: no native signing is possible on mac.

If you're using 33.1.1.68 then yes you'd get that message; native signing for Mac app bundles was added in 33.1.1.533. So if you can download the latest SDK onto the old Mac machine, it should then work?

> I could have created a signed mac.app application on Windows from air, but this gave me a file that was not working

Yes you can't do this, for desktop applications you can only use the target operating system. If you use the native/bundle options on Windows, you get Windows binaries, etc.

> So now you say signing (if I understood you correctly ) - it is not possible to create a signed mac application from air, I can only create it from a swf file

Well yes although actually, if you use the later AIR SDK versions, I think that it also works with creating a bundle from a .air file as well i.e. you should be able to run:
```
adt -package -target bundle -storetype KeychainStore -alias "Developer ID Application" name.app name.air
```

We've just been revisiting this, to also add the ability to sign an existing bundle. Both of these options though, use the native `codesign` command on the machine, which means that there may be challenges if you're using an old macOS version with running the apps on a new macOS version. I don't know whether Apple allow that...

On your new system with Big Sur, it should be possible to get it working, but first you need to un-quarantine the AIR SDK after it's been downloaded/extracted. You can do this via terminal e.g.
```
xattr -r -d com.apple.quarantine air_sdk_folder
```
and after that, it should be possible to then run ADT from the terminal.

One thing we did find is that the new code-signing doesn't appear to work on anything prior to 10.12 (Sierra) even though the binaries are marked such that they should work on 10.10...

Hoe that helps! Let us know how it goes. Sticking with Java 1.8 is a good idea I think...
